test/Microsoft.IdentityModel.Protocols.OpenIdConnect.Tests/ConfigurationManagerTests.cs

```diff
@@ -302,6 +301,41 @@ public void GetConfiguration()
             TestUtilities.AssertFailIfErrors(context);
         }
 
+        // Test checks to make sure that the LastKnownGood (LKG) Configuration lifetime is properly reset at the time
+        // a new LKG is set.
+        [Fact]
+        public void ResetLastKnownGoodLifetime()
+        {
+            TestUtilities.WriteHeader($"{this}.ResetLastKnownGoodLifetime");
+            var context = new CompareContext();
+
+            var validConfig = new OpenIdConnectConfiguration() { TokenEndpoint = Default.Issuer + "oauth/token", Issuer = Default.Issuer };
+            var configurationManager = new MockConfigurationManager<OpenIdConnectConfiguration>(validConfig);
+
+            // set and retrieve config in order to set the first access time
+            configurationManager.LastKnownGoodConfiguration = validConfig;
+            var lkg = configurationManager.LastKnownGoodConfiguration;
+            var lkgConfigFirstUseField = typeof(BaseConfigurationManager).GetField("_lastKnownGoodConfigFirstUse", System.Reflection.BindingFlags.NonPublic | System.Reflection.BindingFlags.Instance);
+            var lkgConfigFirstUse1 = lkgConfigFirstUseField.GetValue(configurationManager as BaseConfigurationManager);
+
+            // set and retrieve config again to reset first access time
+            configurationManager.LastKnownGoodConfiguration = validConfig;
+            lkg = configurationManager.LastKnownGoodConfiguration;
+            var lkgConfigFirstUse2 = lkgConfigFirstUseField.GetValue(configurationManager as BaseConfigurationManager);
+
+            if (lkgConfigFirstUse1 == null)
+                context.AddDiff("Last known good first use time was not properly set for the first configuration.");
+
+            if (lkgConfigFirstUse1 == null)
+                context.AddDiff("Last known good first use time was not properly set for the second configuration.");
+
+            //LKG config first use was not reset when a new configuration was set
+            if (lkgConfigFirstUse1 == lkgConfigFirstUse2)
```

Comment:
question: Thinking from a position of how the test can fail. I don't think this can be ever be an unsuccessful condition (as in actually go to the context.AddDiff line). If we were to comment out lines
`            configurationManager.LastKnownGoodConfiguration = validConfig;
            lkg = configurationManager.LastKnownGoodConfiguration;`

I would think that we're getting the same configuration twice, yet because the references are different this condition would still see them as different.

Is this intended? if not, it might be better to actually change a value in the configuration and compare those values to determine whether a different configuration was obtained or not.

nit: Additionally, imo checking if the configs aren't null would make the test more explicit/descriptive when it fails, as in if one of them is null, getting a failure saying that the config wasn't reset won't add much value.

Reply:
Ah, good catch! I forgot that a DateTime in this language is a value type and not a reference type.

You're right that commenting out those lines does not result in failure, but changing this to make use of Equals() (combined with sleeping for 1 ms) addresses that issue. Seems like even with the use of Equals() the operations happen too fast and so the first use time ends up being the same for both. Not a fan of using Thread.Sleep(), so I'll try to think of other ways to test this particular feature. https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/pull/1723/commits/bd148b61e7f6bd2a1967470437dcb020e6fe14df

The way the code is currently written, we do not check if the value that we're setting for the LKG is the same as the value that is currently set (as it costs more to figure this out than to reset this). The aim of this test isn't to check if a different configuration was obtained, but to make sure that the firstUse property is properly reset when a new LKG is set, as it is used to determine how long an LKG can safely be used for (whether it is valid). Does that make sense?